src/manifest/pfm.rs

```diff
@@ -0,0 +1,497 @@
+/// manifest signature, even if it is parsed long after the TOC has been read
+/// out of flash.
+///
+/// It is not possible to construct a `Toc` directly; it must be parsed out of
+/// a valid [`Pfm`].
+///
+/// [`TocEntry`]: struct.TocEntry.html
+/// [`HashType`]: enum.HashType.html
+/// [`Pfm`]: struct.Pfm.html
+pub struct Toc<'toc> {
+    entries: &'toc [TocEntry],
+    hashes: &'toc [sha256::Digest],
+}
+
+impl<'toc> Toc<'toc> {
+    /// Checks that all invariants of this `Toc` type hold:
+    /// - Every pointer to a hash is in-bounds.
+    /// - Every pointer to a parent is in-bounds.
+    ///
+    /// Returns true if the invariants are upheld.
+    fn verify_invariants(&self) -> bool {
+        for entry in self.entries {
+            if let Some(idx) = entry.hash_idx() {
+                if self.hashes.len() <= idx {
+                    return false;
+                }
+            }
+            if let Some(idx) = entry.parent_idx() {
+                if self.entries.len() <= idx {
+                    return false;
+                }
+            }
+        }
+
+        true
+    }
+
+    /// Returns the number of entries in this `Toc`.
+    pub fn len(&self) -> usize {
+        self.entries.len()
+    }
+
+    /// Returns whether this `Toc` is empty.
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the `i`th entry in this `Toc`.
+    ///
+    /// If there is not `i`th entry, `None` is returned. If the entry specifies
+    /// a hash, it is also returned, otherwise `Some((Some(_), None))` is
+    /// returned.
+    pub fn entry(
+        &self,
+        i: usize,
+    ) -> Option<(&'toc TocEntry, Option<&'toc sha256::Digest>)> {
+        let entry = self.entries.get(i)?;
+        match entry.hash_idx() {
+            // NOTE: This indexing operation cannot panic due to `Toc`'s invariants.
+            Some(idx) => Some((entry, Some(&self.hashes[idx]))),
+            None => Some((entry, None)),
+        }
+    }
+
+    /// Returns an iterator over this `Toc`'s entries and their associated
+    /// hashes, if they specify one.
+    pub fn entries(
+        &self,
+    ) -> impl Iterator<Item = (&'toc TocEntry, Option<&'toc sha256::Digest>)> + '_
+    {
+        // NOTE: The unwrap() below cannot panic, because `i` is always in-bounds.
+        (0..self.entries.len()).map(move |i| self.entry(i).unwrap())
+    }
+
+    /// Returns an iterator over this `Toc`'s entries of a specific type.
+    ///
+    /// The items returned by this iterator also include the index of the entry
+    /// in the overall table.
+    pub fn entries_of_type(
+        &self,
+        ty: ElementType,
+    ) -> impl Iterator<
+        Item = (usize, &'toc TocEntry, Option<&'toc sha256::Digest>),
+    > + '_ {
+        self.entries()
+            .enumerate()
+            .map(|(i, (e, h))| (i, e, h))
+            .filter(move |(_, e, _)| e.element_type() == Some(ty))
```

Comment:
I was going to suggest filtering before mapping, but given it's a type shuffle I don't think it matters, right?

Reply:
One would hope that that map operation is a no-op, given that all three elements have the same size and alignment.